src/js_native_api_v8.cc

```diff
@@ -2359,6 +2382,58 @@ napi_status napi_create_external(napi_env env,
   return napi_clear_last_error(env);
 }
 
+NAPI_EXTERN napi_status napi_type_tag_object(napi_env env,
+                                             napi_value object,
+                                             const napi_type_tag* type_tag) {
+  NAPI_PREAMBLE(env);
+  v8::Local<v8::Context> context = env->context();
+  v8::Local<v8::Object> obj;
+  CHECK_TO_OBJECT_WITH_PREAMBLE(env, context, obj, object);
+
+  auto key = NAPI_PRIVATE_KEY(context, type_tag);
+  auto maybe_has = obj->HasPrivate(context, key);
+  CHECK_MAYBE_NOTHING_WITH_PREAMBLE(env, maybe_has, napi_generic_failure);
+  RETURN_STATUS_IF_FALSE_WITH_PREAMBLE(env,
+                                       !maybe_has.FromJust(),
+                                       napi_invalid_arg);
+
+  auto ref = v8impl::TypeTagReference::New(env, obj, type_tag);
+  auto tag = v8::External::New(env->isolate, ref);
```

Comment:
This creates a JS object, a C++ object and a weak handle (and it does not re-use them when using `napi_type_tag_object()` on multiple objects). Using a BigInt to store the type tag value directly would probably have both noticeably lower overhead and a simpler implementation.